### PR TITLE
Leaves - Nicky

### PR DESCRIPTION
**Pull request by njch5**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I created a custom top_ten method in my Work model where it would return ten works for each category in descending order by number of votes.
Describe how you approached testing that model method. What edge cases did you come up with? | If there are no works, then an empty array would be returned. Another edge case is if there are less than ten works, the method will still return a list and won't be nil.
What are `session` and `flash`? What is the difference between them? | Both are special type of hashes that persist data between requests. Typically, session is used to keep track of a user's data when logged into a site, its data is stored indefinitely. Whereas, flash sends one-time messages that only persist from one request to the next.
What was one thing that you gained more clarity on through this assignment? | I gained clarity on testing the model classes in Rails and coming up with edge cases.
What is the Heroku URL of your deployed application? | https://media-ranker-nc.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | For testing the models' relations, it would be helpful to provide some example test cases for a "has_many" association since the example we did in class was for a "belongs_to" association.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      session[:user_id] = user.id
+      flash[:status] = :success
+      flash[:message] = "Successfully created new user #{username} with ID #{user.id}"
+    end
+
+    redirect_to root_path
+  end
+
+  def current
+    @current_user = User.find_by(id: session[:user_id])
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:38]: Finding a logged in user is good functionality to put in a controller filter because it may be used multiple places in the future.

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    before do
+      users(:test).save
+      @user = User.first
+    end
+    it "has a zero to many relationship with votes" do
+      expect(@user.votes.count).must_be :>=, 0
+    end
+
+    it "can have many votes" do
+      @user.votes.each do |vote|
```
Comment (beccaelenzil) [inline test/models/user_test.rb:30]: You never make any `votes` for this `user`. Consider: why doesn't this test fail?

File: test/models/vote_test.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    User.create!(username: "new user")
+  }
+  let (:new_vote) {
+    Vote.new(work_id: work.id, user_id: user.id)
+  }
+  describe "relationships" do
+    describe "relations with Work" do
+      it "belongs to a work and can set the work_id through work" do
+        new_vote.work = work
+        expect(new_vote.work).must_be_instance_of Work
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:18]: This test simply confirms an assignment.

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        expect(new_vote.work_id).must_equal work.id
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:19]: this test tests the relationship, yay!

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | no flash message
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | ye
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | no votes.yml 
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | See comment -- there are not votes for the user you are testing so you can't test this relation. Either make the votes in your test, or use a votes.yml file.
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | As with users, you haven't added votes so you can't test this relationship here.
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good work overall! Your implementation matches the demo site very closely, and the learning goals for this assignment were definitely met. One place for growth is in your tests of relationships. You have not fully tested the relationship between votes and users/works. One step to achieving this full testing would be to create a `votes.yml` file with vote fixtures that include a `user` and `work` from your `user` and `work` fixtures.

File: app/views/homepages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        <span class="spotlight__header-prefix">Media Spotlight: </span>
+        <%= link_to Work.top_media.title, work_path(Work.top_media.id) %> by <%= Work.top_media.creator %>
+      </h2>
+
+      <p class="spotlight__description"><%= Work.top_media.votes.count %> votes - Description: <%= Work.top_media.description %> </p>
+  </section>
+
+  <hr class="root__hr">
+
+  <section class="top-ten__container">
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:13]: Consider using a partial view to DRY up this repeated code.
